// did-jwt-rsa/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RSASigner = exports.DID = exports.toEthereumAddress = exports.createJWS = exports.verifyJWS = exports.decodeJWT = exports.createJWT = exports.verifyJWT = exports.decryptJWE = exports.createJWE = void 0;
const JWT_1 = require("./JWT");
Object.defineProperty(exports, "verifyJWT", { enumerable: true, get: function () { return JWT_1.verifyJWT; } });
Object.defineProperty(exports, "createJWT", { enumerable: true, get: function () { return JWT_1.createJWT; } });
Object.defineProperty(exports, "decodeJWT", { enumerable: true, get: function () { return JWT_1.decodeJWT; } });
Object.defineProperty(exports, "verifyJWS", { enumerable: true, get: function () { return JWT_1.verifyJWS; } });
Object.defineProperty(exports, "createJWS", { enumerable: true, get: function () { return JWT_1.createJWS; } });
const Digest_1 = require("./Digest");
Object.defineProperty(exports, "toEthereumAddress", { enumerable: true, get: function () { return Digest_1.toEthereumAddress; } });
const RSASigner_1 = require("./signers/RSASigner");
Object.defineProperty(exports, "RSASigner", { enumerable: true, get: function () { return RSASigner_1.RSASigner; } });
const dids_1 = require("./dids");
Object.defineProperty(exports, "DID", { enumerable: true, get: function () { return dids_1.DID; } });
var JWE_1 = require("./JWE");
Object.defineProperty(exports, "createJWE", { enumerable: true, get: function () { return JWE_1.createJWE; } });
Object.defineProperty(exports, "decryptJWE", { enumerable: true, get: function () { return JWE_1.decryptJWE; } });
